content/unroll-default-arguments.md

```diff
@@ -0,0 +1,817 @@
+   versions of the upstream code. If we can come up with some implementation that works, we can
+   lift this restriction later, but for now I have not managed to do so and so this restriction
+   stays.
+
+### Challenges of Non-Final Methods and Overriding
+
+To elaborate a bit on the issues with non-final methods and overriding, consider the following 
+case with four classes, `Upstream`, `Downstream`, `Main1` and `Main2`, each of which is compiled
+against different versions of each other (hence the varying number of parameters for `foo`):
+
+```scala
+class Upstream{ // V2
+   def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0) = s + n + b + l
+}
+```
+
+```scala
+class Downstream extends Upstream{ // compiled against Upstream V1
+   override def foo(s: String, n: Int = 1) = super.foo(s, n) + s + n
+}
+```
+
+```scala
+object Main1 { // compiled against Upstream V2
+   def main(args: Array[String]): Unit = {
+      new Downstream().foo("hello", 123, false, 456L)
+   }
+}
+```
+
+```scala
+object Main2 { // compiled against Upstream V1
+   def main(args: Array[String]): Unit = {
+      new Downstream().foo("hello", 123)
+   }
+}
+```
+
+
+The challenge here is: how do we make sure that `Main1` and `Main2`, who call
+`new Downstream().foo`, correctly pick up the version of `def foo` that is 
+provided by `Downstream`? 
+
+With the current implementation, the `override def foo` inside `Downstream` would only
+override one of `Upstream`'s synthetic forwarders, but would not override the actual
+primary implementation. As a result, we would see `Main1` calling the implementation
+of `foo` from `Upstream`, while `Main2` calls the implementation of `foo` from 
+`Downstream`. So even though both `Main1` and `Main2` have the same 
+`Upstream` and `Downstream` code on the classpath, they end up calling different
+implementations based on what they were compiled against.
+
+We cannot perform the method search and dispatch _within_ the `def foo` methods,
+because it depends on exactly _how_ `foo` is called: the `InvokeVirtual` call from
+`Main1` is meant to resolve to `Downstream#foo`, while the `InvokeSpecial` call
+from `Downstream#foo`'s `super.foo` is meant to resolve to `Upstream#foo`. But a
+method implementation cannot know how it was called, and thus it is impossible
+for `def foo` to forward the call to the right place.
+
+This issue only arises in the presence of version skew between `Upstream` and 
+`Downstream` code, but that scenario is precisely where `@unroll` is meant to
+provide benefits. Thus, unless we can find some solution, we cannot properly support
+virtual methods and overrides in `@unroll`.
+
+It may be possible to loosen this restriction to also allow abstract methods that
+are implemented only once by a final method. See the section about 
+[Abstract Methods](#abstract-methods) for details.
+
+## Major Alternatives
+
+The major alternatives to `@unroll` are listed below:
```

Comment:
I would add an implementation with macro annotations to that list. Macro annotations are about to land as a SIP. We should evaluate whether and how they could implement this behavior. If they can do it, we should go down that road, I think. Then the discussion would be where the new macro should live. I would argue that it should go into the standard library in that case.

Reply:
Yes that's definitely worth a consideration. I will need to wait for the macro annotations proposal to appear before I can really say, but if we can implement this in a macro annotation rather than hardcoded in the compiler that would definitely be preferable.

Reply:
We can propose it as a test case for the implementers of macro annotations /cc @nicolasstucki @hamzaremmal.

Reply:
I do think we do not need to block on macro annotations before reviewing this proposal. 

Whether `@unroll` ends up being baked into the compiler, implemented as a compiler plugin (as it is today), or implemented using macro annotations, the feature is still the same. It's just a matter of what line you need in your build file to enable it.

e.g., we could decide that we want `@unroll` as a feature, leave it a compiler plugin while macro annotations are figured out (possibly under the `github.com/scala` org?), and eventually fold it into the standard library as a macro annotation when those are ready. The Scala 2 implementation will need to be done differently regardless, either baked-in or remaining as a compiler plugin.

Reply:
Got it. 

alAn earlier version of the prototype had an `@unroll("x")` annotation on the method, where the `"x"` string parameter specified the parameter name from which to begin unrolling.

We could also do two annotations: `@unroll` on the method, and then `@unroll2` or something on the starting parameter.

I don't have a strong opinion between these options. The current API was chosen as the most minimal while being able to provide the flexibility necessary, but the others are fine too IMO (if marginally more verbose)